KIC Proxy Request guides (#879)
Kong/developer.konghq.com: app/_how-tos/kic-proxy-tcp-port.md

---
title: Proxy TCP traffic by port
description: "Route TCP requests to services in your cluster based on the incoming port using `TCPRoute` or `TCPIngress`"
content_type: how_to
related_resources:
  - text: All KIC documentation
    url: /index/kubernetes-ingress-controller/

permalink: /kubernetes-ingress-controller/routing/tcp-by-port/
breadcrumbs:
  - /kubernetes-ingress-controller/
  - index: kubernetes-ingress-controller
    section: Routing

products:
  - kic

works_on:
  - on-prem
  - konnect

entities: []

tldr:
  q: How do I route TCP traffic with {{ site.kic_product_name }}?

>   a: Create a `TCPRoute` or `TCPIngress` resource, which will then be converted in to a {{ site.base_gateway }} Service and Route

Reviewer: ```suggestion
  a: Create a `TCPRoute` or `TCPIngress` resource, which will then be converted in to a [{{ site.base_gateway }} Service](/gateway/entities/service/) and [Route](/gateway/entities/route/).
```

prereqs:
  kubernetes:
    gateway_api: experimental
  entities:
    services:
      - echo-service

cleanup:
  inline:
    - title: Uninstall KIC from your cluster
      include_content: cleanup/products/kic
      icon_url: /assets/icons/kic.svg
---

{% include /k8s/configure-tcp-listen.md plaintext=true tls=false %}

## Route TCP traffic

> To expose the service to the outside world, create a TCPRoute resource for Gateway APIs or a TCPIngress resource for Ingress.

Reviewer: ```suggestion
To publicly expose the service, create a `TCPRoute` resource for Gateway APIs or a `TCPIngress` resource for Ingress.
```
Not sure if that's the right wording.

{% navtabs api %}
{% navtab "Gateway API" %}

To reconcile the `TCPRoute`, configure an additional TCP listener on your `Gateway` resource:

```bash
kubectl patch -n kong --type=json gateway kong -p='[
    {
        "op":"add",
        "path":"/spec/listeners/-",
        "value":{
            "name":"stream9000",
            "port":9000,
            "protocol":"TCP"
        }
    }
]'
```

Next, create a `TCPRoute`:

```bash
echo "apiVersion: gateway.networking.k8s.io/v1alpha2
kind: TCPRoute
metadata:
  name: echo-plaintext
  namespace: kong
spec:
  parentRefs:
  - name: kong
    sectionName: stream9000
  rules:
  - backendRefs:
    - name: echo
      port: 1025
" | kubectl apply -f -
```

{% endnavtab %}
{% navtab "Ingress" %}

```bash
echo "apiVersion: configuration.konghq.com/v1beta1
kind: TCPIngress
metadata:
  name: echo-plaintext
  namespace: kong
  annotations:
    kubernetes.io/ingress.class: kong
spec:
  rules:
  - port: 9000
    backend:
      serviceName: echo
      servicePort: 1025
" | kubectl apply -f -
```

{% endnavtab %}
{% endnavtabs %}

This configuration instructs {{site.base_gateway}} to forward all traffic it

> receives on port 9000 to `echo` service on port 1025.

Reviewer: ```suggestion
receives on port 9000 to the `echo` service on port 1025.
```

## Validate your configuration

You can now test your route using `telnet`.

```shell
$ telnet $PROXY_IP 9000
```

After you connect, type some text that you want as a response from the echo Service.

```
Trying 192.0.2.3...
Connected to 192.0.2.3.
Escape character is '^]'.
Welcome, you are connected to node gke-harry-k8s-dev-pool-1-e9ebab5e-c4gw.
Running on Pod echo-844545646c-gvmkd.
In namespace default.
With IP address 192.0.2.7.
This text will be echoed back.
This text will be echoed back.
^]
telnet> Connection closed.
```

> To exit, press `ctrl+]` then `ctrl+d`.

Reviewer: ```suggestion
To exit, press `Control+]` then `Control+D`.
```

Author: `ctrl` is the common formatting I believe
